### DISCUSSION TITLE
JanusGraph not working remotely, but working when used through JanusGraph Core

### DISCUSSION THREAD

**Discussion by sharma-kunal** [JanusGraph/janusgraph#4002] [Q&A]
I am creating a graph using Janusgraph with Cassandra as the backend and elastic for indexing. And I am stuck with a strange issue, 

I create my graph like this, create a traversal, add some vertices and edges in it. 
```
graph = JanusGraphFactory.build()
                .set("graph.set-vertex-id", true)
                .set("query.force-index", true)
                .set("storage.backend", "cql")
                .set("storage.cql.keyspace", "test")
                .set("storage.hostname", "<Cassandra IP>")
                .set("storage.port", "9042")
                .set("index.test.backend", "elasticsearch")
                .set("index.test.index-name", "test")
                .set("index.test.hostname", "<Elastic IP>")
                .set("index.test.port", "9200")
                .set("index.test.elasticsearch.http.auth.type", "BASIC")
                .set("index.test.elasticsearch.http.auth.basic.username", "<Username>")
                .set("index.test.elasticsearch.http.auth.basic.password", "<Password>")
                .set("graph.replace-instance-if-exists", true)
                .set("tx.log-tx", true)
                .set("tx.max-commit-time", 10000)
                .open();
```

After which I run a simple query,

```
g.V().hasLabel("seeker").has("p_uuid", "mno789").next()
```

and it works fine. It returns me the vertex

but when I try to do it from python, 

```
connection = DriverRemoteConnection(url='ws://<Janus Server IP>:8182/gremlin', traversal_source='test')
g = traversal().withRemote(connection)
```

I get an empty result. 

This is my janus server .properties file

```
graph.set-vertex-id=true
storage.backend=cql
storage.cql.keyspace=test
storage.hostname=<hostname ip>
storage.port=9042
index.test.backend=elasticsearch
index.test.index-name=test
index.test.hostname=<elastic ip>
index.test.port=9200
index.test.elasticsearch.http.auth.type=BASIC
index.test.elasticsearch.http.auth.basic.username=<username>
index.test.elasticsearch.http.auth.basic.password=<password>
graph.replace-instance-if-exists=true
graph.force-index=true
tx.log-tx=true
tx.max-commit-time=10000
```

I am unable to figure out the cause for this ? Is there any way you would approach to debug the issue ? I have gone through the logs but could not find anything relevant. It would be really helpful if I could get some guidance to resolve this issue. 

One point to notice is, I create two graphs like this with same configs, and one is working fine, but the other is not. There is no difference in their config, still one is not working.

**Comment by vtslab** [comment]
This sounds like a commit issue. The embbeded JanusGraph from the first code block and the JanusGraph inside JanusGraph server are (almost) independent JanusGraph instances that use the same storage keyspace. The instances only know about changes in the graph when the transaction with these changes is committed. For remote connections transactions are committed on each request. When using embedded Janusgraph with janusgraph-core, you have to issue an explicit g.tx().commit() to persist the changes.

**Comment by sharma-kunal** [comment]
I have two graphs, education and experience, and as you stated I need to commit the changes, I do that as well

```
educationGraphTraversalSource.tx().commit();
experienceGraphTraversalSource.tx().commit();
educationGraphTraversalSource.tx().close();
experienceGraphTraversalSource.tx().close();
```

and changes are reflected in the education graph, but in the experience graph, they are not. 

Now, since for experience graph it is working fine when I use it from the first code block, that means data is committed in cassandra, then this issue is related to Janusgraph ? I mean, which logs should I be focusing more into ?

**Comment by sharma-kunal** [comment]
I tried removing the `query.force-index` flag, and it is working fine now. So, does this mean that the data was not present in index ? I am using both composite and mixed index (elastic). And as I checked from my kibana dashboard, data is present in Elastic, but when I run the search query,

```
g.V().hasLabel("seeker").has("p_uuid", "mno789").toList()
```

it will use composite index and not mixed. So, is it possible that data is added in cassandra, and mixed index but not in comosite index ? and if yes, how can I check if this is the case ? Whether data is present in composite index or not ?

This is the output of the .profile() query

```
[{'dur': 9.797251, 'metrics': [{'dur': 9.797251, 'counts': {'traverserCount': 1, 'elementCount': 1}, 'name': 'JanusGraphStep([],
[~label.eq(seeker), p_uuid.eq(684e082d1471)])', 'annotations': {'percentDur': 100.0}, 'id': '4.0.0()', 'metrics': 
[{'dur': 1.857527, 'counts': {}, 'name': 'constructGraphCentricQuery', 'id': '4.0.0().constructGraphCentricQuery_0'}, 
{'dur': 3.165153, 'counts': {}, 'name': 'GraphCentricQuery', 'annotations': {'condition': '(~label = seeker AND p_uuid = 
684e082d1471)', 'orders': '[]', 'isFitted': 'true', 'isOrdered': 'true', 'query': 'multiKSQ[1]', 'index': 'seeker'}, 'id': 
'4.0.0().GraphCentricQuery_1', 'metrics': [{'dur': 1.645895, 'counts': {'elementCount': 1}, 'name': 'backend-query', 
'annotations': {'query': 'seeker:multiKSQ[1]'}, 'id': '4.0.0().GraphCentricQuery_1.backend-query_0'}]}]}]}]
```

and it states that index was used and that too conmposite index. So why if the flag `query.force-index` is set to `true`, it is not working ? Any suggestion, am I missing anything here ?
